File: server/comfyfed_server/scheduler.py
"""Phase 3.3 §2.4-§2.5: 成本模型與整體配對。

全部是純函數：不碰 DB、不讀時鐘（`now` 一律由呼叫端傳進來）。呼叫端
（`dispatch.assign_jobs`）負責把 DB 列翻成這裡的 `JobCandidate` /
`WorkerCandidate` / `PairVerdict`，再把結果翻回去做原子 claim。

兩棧逐行對照 `cloud/src/core/scheduler.ts`；任何一邊改了算式，另一邊必須
同一個 commit 一起改，共用 fixture `cloud/test/fixtures/scheduler_cases.json`
就是用來釘住這件事的。
"""

from __future__ import annotations

import math
from dataclasses import dataclass
from datetime import datetime
from typing import Optional, Sequence

from . import assess

LOAD_SEC_PER_GB = 1.5
FETCH_BYTES_PER_SEC = 50e6
STARVE_SECONDS = 300
# 等待超過 STARVE_SECONDS 再多減這麼多，確保只要有合格 worker 一定本 tick 派出。
STARVE_BONUS = 1e8
AGE_WEIGHT = 1.0
BIG = 1e9
WARN_PENALTY = 1_000_000.0
LIGHT_BACKEND_PENALTY = 60.0
LIGHT_VRAM_WEIGHT = 5.0

_WEAK_BACKENDS = ("mps", "cpu")

# Hungarian 內部用的「禁止」哨兵。演算法裡會做 `a[i][j] - u[i] - v[j]` 這種
# 減法，真的塞 `inf` 會冒出 `inf - inf = NaN` 把整個 potential 弄壞，所以禁止
# 的格子換成一個大到絕不會被選中、但仍是有限值的數；選出來之後再回頭對照
# 原始矩陣把禁止格剔掉。
_FORBIDDEN_SENTINEL = 1e18


@dataclass(frozen=True)
class JobCandidate:
    """派工決策需要的 job 欄位，和 ORM 脫鉤。"""

    job_id: str
    signature: Optional[str]
    created_at: datetime
    # 無 required_models 且無 est_vram_gb -- 見 dispatch 的 `is_light`。
    is_light: bool
    required_models: tuple[str, ...]


@dataclass(frozen=True)
class WorkerCandidate:
    worker_id: str
    name: str
    backend: str
    free_vram_gb: float
    # 這台 worker 最近一次被指派的 job 的 required_models（§2.2）。
    warm_models: tuple[str, ...]
    # `assess.model_inventory(worker)` 的結果，用來查模型大小。
    inventory: tuple[dict, ...]


@dataclass(frozen=True)
class PairVerdict:
    """`assess.verdict` 對這一對 (job, worker) 的判定，壓成排程需要的三個值。"""

    kind: str  # "eligible" | "eligible_after_fetch" | "ineligible"
    has_warnings: bool
    total_fetch_bytes: int


def _finite(value) -> bool:
    return (
        isinstance(value, (int, float))
        and not isinstance(value, bool)
        and math.isfinite(float(value))
    )


def load_seconds(job: JobCandidate, worker: WorkerCandidate) -> float:
    """§2.4：這台 worker 還沒熱起來的模型，預估要花多久載進 VRAM。

    大小來自 worker 自己的 `model_inventory`（用 `assess.find_model` 比對，
    因為工作流的模型名和 inventory 路徑根目錄不同）；查不到大小視為 0，
    寧可低估也不要憑空捏一個數字進成本。
    """
    warm = set(worker.warm_models)
    total_gb = 0.0
    for name in job.required_models:
        if name in warm:
            continue
        _found, size_gb = assess.find_model(list(worker.inventory), name)
        if size_gb is not None:
            total_gb += float(size_gb)
    return total_gb * LOAD_SEC_PER_GB


def fetch_seconds(pair: PairVerdict) -> float:
    """§2.4：只有 `eligible_after_fetch` 才會 > 0（頻寬用常數，不量測）。"""
    if pair.kind != "eligible_after_fetch":
        return 0.0
    return float(pair.total_fetch_bytes) / FETCH_BYTES_PER_SEC


def light_penalty(job: JobCandidate, worker: WorkerCandidate) -> float:
    """§2.4：零模型工作落在大卡上要付的代價 -- 把大卡留給真的需要的工作。"""
    if not job.is_light:
        return 0.0
    backend_penalty = 0.0 if worker.backend in _WEAK_BACKENDS else LIGHT_BACKEND_PENALTY
    return backend_penalty + worker.free_vram_gb * LIGHT_VRAM_WEIGHT


def _legacy_tiebreak(job: JobCandidate, worker: WorkerCandidate) -> float:
    """只用來打破完全相等的舊排序鍵：輕工作偏小卡、重工作偏大卡。

    量級刻意做到小於 1 秒（重工作那側甚至小於 1 毫秒），所以它永遠不會蓋過
    任何一項真正的成本，只在兩邊其他每一項都一模一樣時說話。
    """
    if job.is_light:
        return worker.free_vram_gb / 1000.0
    return (1000.0 - worker.free_vram_gb) / 1e6


def cost(
    job: JobCandidate,
    worker: WorkerCandidate,
    pair: PairVerdict,
    predicted_seconds: float,
    tier1_exists: bool,
) -> float:
    """§2.4 的 `cost(j, w)`；不可用一律 `math.inf`。

    `tier1_exists` 是「這個 job 至少有一個 `eligible` 候選」-- 有的話，
    `eligible_after_fetch` 的候選對這個 job 一律視為不可用，保住「已經有模型
    的一定贏過要下載的」這個既有語意。

    任何一項算出 NaN 或負數也回 `inf`：成本本來就不該是負的，出現代表上游
    算錯了，寧可不派也不要讓一個壞值贏過所有合理選擇（spec §5）。
    """
    if pair.kind not in ("eligible", "eligible_after_fetch"):
        return math.inf
    if pair.kind == "eligible_after_fetch" and tier1_exists:
        return math.inf

    components = (
        predicted_seconds,
        load_seconds(job, worker),
        fetch_seconds(pair),
        light_penalty(job, worker),
        WARN_PENALTY if pair.has_warnings else 0.0,
        _legacy_tiebreak(job, worker),
    )
    total = 0.0
    for component in components:
        if not _finite(component) or component < 0:
            return math.inf
        total += float(component)
    return total


def objective(cost_value: float, wait_seconds: float) -> float:
    """§2.5 第 4 條的目標函數項。

    減掉 `BIG` 是為了讓「多派一件」永遠優於「少派一件」：任何一個真實配對的
    目標值都遠低於補零的那些格子，所以 Hungarian 在不違反禁止格的前提下一定
    會盡量多配對。
    """
    if not _finite(cost_value):
        return math.inf
    wait = wait_seconds if _finite(wait_seconds) and wait_seconds > 0 else 0.0
    value = cost_value - AGE_WEIGHT * wait - BIG
    if wait >= STARVE_SECONDS:
        value -= STARVE_BONUS
    return value


def build_matrix(
    jobs: Sequence[JobCandidate],
    workers: Sequence[WorkerCandidate],
    pairs: dict[tuple[str, str], PairVerdict],
    predictions: dict[tuple[str, str], float],
    now: datetime,
) -> list[list[float]]:
    """`matrix[i][j]` = job i 派給 worker j 的目標值；不可用為 `math.inf`。

    `pairs` / `predictions` 以 `(job_id, worker_id)` 為鍵；缺鍵視為不可用／
    使用預設 60 秒（`predictions` 缺鍵只可能是呼叫端漏算，不該讓整個 tick
    掛掉）。
    """
    tier1 = {
        job.job_id: any(
            pairs.get((job.job_id, worker.worker_id), PairVerdict("ineligible", False, 0)).kind
            == "eligible"
            for worker in workers
        )
        for job in jobs
    }

    matrix: list[list[float]] = []
    for job in jobs:
        wait_seconds = (now - job.created_at).total_seconds()
        row: list[float] = []
        for worker in workers:
            pair = pairs.get((job.job_id, worker.worker_id))
            if pair is None:
                row.append(math.inf)
                continue
            predicted = predictions.get((job.job_id, worker.worker_id), 60.0)
            row.append(
                objective(cost(job, worker, pair, predicted, tier1[job.job_id]), wait_seconds)
            )
        matrix.append(row)
    return matrix


def solve(matrix: list[list[float]]) -> list[tuple[int, int]]:
    """最小成本配對（Kuhn–Munkres / Hungarian，O(n³)）。

    長方形輸入會內部補 0 成方陣。非有限值（`inf`、`-inf`、`NaN`）視為禁止，
    最後回傳時剔除；有限負值完全合法（目標函數刻意是負的）。回傳依 row 排序
    的 `(row, col)`，不含補零的行列。

    決定性：在固定的輸入順序下，內層挑 `delta` 用的是嚴格小於，所以平手時
    永遠選欄位索引最小的那個；TS 版逐行相同，因此兩棧同一個矩陣得到同一組
    配對。呼叫端要先把 jobs 依 `(created_at, job_id)`、workers 依
    `(name, worker_id)` 排好，這個保證才有意義。
    """
    rows_n = len(matrix)
    cols_n = max((len(row) for row in matrix), default=0)
    n = max(rows_n, cols_n)
    if n == 0:
        return []

    # 1-indexed 工作矩陣。
    a = [[0.0] * (n + 1) for _ in range(n + 1)]
    for i in range(n):
        for j in range(n):
            value = 0.0
            if i < rows_n and j < len(matrix[i]):
                raw = matrix[i][j]
                value = float(raw) if _finite(raw) else _FORBIDDEN_SENTINEL
            a[i + 1][j + 1] = value

    inf = math.inf
    u = [0.0] * (n + 1)
    v = [0.0] * (n + 1)
    p = [0] * (n + 1)
    way = [0] * (n + 1)

    for i in range(1, n + 1):
        p[0] = i
        j0 = 0
        minv = [inf] * (n + 1)
        used = [False] * (n + 1)
        while True:
            used[j0] = True
            i0 = p[j0]
            delta = inf
            j1 = 0
            for j in range(1, n + 1):
                if used[j]:
                    continue
                cur = a[i0][j] - u[i0] - v[j]
                if cur < minv[j]:
                    minv[j] = cur
                    way[j] = j0
                if minv[j] < delta:
                    delta = minv[j]
                    j1 = j
            for j in range(0, n + 1):
                if used[j]:
                    u[p[j]] += delta
                    v[j] -= delta
                else:
                    minv[j] -= delta
            j0 = j1
            if p[j0] == 0:
                break
        while True:
            j1 = way[j0]
            p[j0] = p[j1]
            j0 = j1
            if j0 == 0:
                break

    result: list[tuple[int, int]] = []
    for j in range(1, n + 1):
        i = p[j]
        if i == 0:
            continue
        row, col = i - 1, j - 1
        if row >= rows_n or col >= cols_n or col >= len(matrix[row]):
            continue
        if not _finite(matrix[row][col]):
            continue
        result.append((row, col))
    result.sort()
    return result


def match(
    jobs: Sequence[JobCandidate],
    workers: Sequence[WorkerCandidate],
    pairs: dict[tuple[str, str], PairVerdict],
    predictions: dict[tuple[str, str], float],
    now: datetime,
) -> list[tuple[int, int]]:
    """`build_matrix` + `solve` 的便利包裝：回傳 `(job index, worker index)`。"""
    return solve(build_matrix(jobs, workers, pairs, predictions, now))

File: cloud/src/core/scheduler.ts
/**
 * Phase 3.3 §2.4-§2.5: 成本模型與整體配對。
 * 自原 Python 伺服器移植（2026-09）；本檔現為唯一實作。全部是純函數，不碰
 * D1、不讀時鐘（`now` 由呼叫端傳）。改了算式，`cloud/test/fixtures/
 * scheduler_cases.json`（釘住算式的黃金 fixture）必須同一個 commit 一起改。
 */

import { findModel } from "./assess";
import type { ModelInventoryEntry } from "../db/queries";

export const LOAD_SEC_PER_GB = 1.5;
export const FETCH_BYTES_PER_SEC = 50e6;
export const STARVE_SECONDS = 300;
/** 等待超過 STARVE_SECONDS 再多減這麼多，確保只要有合格 worker 一定本 tick 派出。 */
export const STARVE_BONUS = 1e8;
export const AGE_WEIGHT = 1.0;
export const BIG = 1e9;
export const WARN_PENALTY = 1_000_000.0;
export const LIGHT_BACKEND_PENALTY = 60.0;
export const LIGHT_VRAM_WEIGHT = 5.0;

/** 2026-09-23 cross-backend dispatch (spec: docs/superpowers/specs/
 * 2026-09-23-cross-backend-dispatch-design.md) -- ports
 * `scheduler.BACKEND_SPEED_PRIOR`: a worker that has never completed a job
 * has no learned `speed_index`; until it does, its backend supplies a prior
 * so the first predictions are in the right order of magnitude (an Apple M4
 * Pro measured ~1/8 of an RTX 5080 on fp8 Chroma, and worse on emulated int8
 * video). 1.0 = "fleet average NVIDIA card". */
export const BACKEND_SPEED_PRIOR: Record<string, number> = { cuda: 1.0, rocm: 0.7, mps: 0.12, cpu: 0.03 };
export const DEFAULT_SPEED_PRIOR = 0.5;

/** Wait-or-run-now (§3): a job is HELD off an idle worker when a currently
 * busy worker is expected to finish it sooner even after waiting for it.
 * The margin absorbs estimate error -- we only hold when waiting is clearly
 * better, never on a coin flip. Ports `scheduler.HOLD_MARGIN_*`. */
export const HOLD_MARGIN_RATIO = 1.25;
export const HOLD_MARGIN_SECONDS = 30.0;
/** A running job that has already taken longer than this multiple of its own
 * prediction has an unknown remaining time: nobody is held for that worker. */
export const OVERDUE_FACTOR = 2.0;
/** Remaining time never drops below this fraction of the prediction, so a
 * worker "about to finish" for the last ten minutes cannot hold a job with
 * an ETA of 0. */
export const MIN_REMAINING_FRACTION = 0.1;
export const HELD_KIND = "held";

const WEAK_BACKENDS = ["mps", "cpu"];

export interface JobCandidate {
  jobId: string;
  signature: string | null;
  createdAt: Date;
  /** 無 required_models 且無 est_vram_gb。 */
  isLight: boolean;
  requiredModels: string[];
}

export interface WorkerCandidate {
  workerId: string;
  name: string;
  backend: string;
  freeVramGb: number;
  /** 最近一次被指派的 job 的 required_models（§2.2）。 */
  warmModels: string[];
  inventory: ModelInventoryEntry[];
}

export interface PairVerdict {
  kind: string; // "eligible" | "eligible_after_fetch" | "ineligible"
  hasWarnings: boolean;
  totalFetchBytes: number;
}

/** `pairs` / `predictions` 的鍵。TS 沒有 tuple key，用一個不可能出現在 id 裡
 * 的 NUL 分隔字串取代 Python 的 `(job_id, worker_id)`。 */
export function pairKey(jobId: string, workerId: string): string {
  return `${jobId}\u0000${workerId}`;
}

function finite(value: unknown): value is number {
  return typeof value === "number" && Number.isFinite(value);
}

/** §2.4 -- ports `scheduler.load_seconds`. */
export function loadSeconds(job: JobCandidate, worker: WorkerCandidate): number {
  const warm = new Set(worker.warmModels);
  let totalGb = 0;
  for (const name of job.requiredModels) {
    if (warm.has(name)) continue;
    const [, sizeGb] = findModel(worker.inventory, name);
    if (sizeGb !== null) totalGb += sizeGb;
  }
  return totalGb * LOAD_SEC_PER_GB;
}

/** §2.4 -- ports `scheduler.fetch_seconds`. */
export function fetchSeconds(pair: PairVerdict): number {
  if (pair.kind !== "eligible_after_fetch") return 0;
  return pair.totalFetchBytes / FETCH_BYTES_PER_SEC;
}

/** §2.4 -- ports `scheduler.light_penalty`. */
export function lightPenalty(job: JobCandidate, worker: WorkerCandidate): number {
  if (!job.isLight) return 0;
  const backendPenalty = WEAK_BACKENDS.includes(worker.backend) ? 0 : LIGHT_BACKEND_PENALTY;
  return backendPenalty + worker.freeVramGb * LIGHT_VRAM_WEIGHT;
}

/** 只用來打破完全相等的舊排序鍵 -- ports `scheduler._legacy_tiebreak`. */
function legacyTiebreak(job: JobCandidate, worker: WorkerCandidate): number {
  if (job.isLight) return worker.freeVramGb / 1000;
  return (1000 - worker.freeVramGb) / 1e6;
}

/** §2.4 的 `cost(j, w)` -- ports `scheduler.cost`；不可用一律 Infinity。 */
export function cost(
  job: JobCandidate,
  worker: WorkerCandidate,
  pair: PairVerdict,
  predictedSeconds: number,
  tier1Exists: boolean
): number {
  if (pair.kind !== "eligible" && pair.kind !== "eligible_after_fetch") return Infinity;
  if (pair.kind === "eligible_after_fetch" && tier1Exists) return Infinity;

  const components = [
    predictedSeconds,
    loadSeconds(job, worker),
    fetchSeconds(pair),
    lightPenalty(job, worker),
    pair.hasWarnings ? WARN_PENALTY : 0,
    legacyTiebreak(job, worker),
  ];
  let total = 0;
  for (const component of components) {
    if (!finite(component) || component < 0) return Infinity;
    total += component;
  }
  return total;
}

/** §2.5 第 4 條 -- ports `scheduler.objective`. */
export function objective(costValue: number, waitSeconds: number): number {
  if (!finite(costValue)) return Infinity;
  const wait = finite(waitSeconds) && waitSeconds > 0 ? waitSeconds : 0;
  let value = costValue - AGE_WEIGHT * wait - BIG;
  if (wait >= STARVE_SECONDS) value -= STARVE_BONUS;
  return value;
}

/** The `speed_index` a worker with no completed job is assumed to have --
 * ports `scheduler.speed_prior`. */
export function speedPrior(backend: string): number {
  return BACKEND_SPEED_PRIOR[backend || ""] ?? DEFAULT_SPEED_PRIOR;
}

/** §3.1 -- ports `scheduler.remaining_seconds`: how much longer a running
 * job is expected to take, or null when that is unknowable (no usable
 * prediction, or already overdue past `OVERDUE_FACTOR` -- the estimate was
 * wrong and must not anchor a hold). */
export function remainingSeconds(predicted: number, elapsed: number): number | null {
  if (!finite(predicted) || predicted <= 0 || !finite(elapsed) || elapsed < 0) return null;
  if (elapsed > predicted * OVERDUE_FACTOR) return null;
  return Math.max(predicted - elapsed, predicted * MIN_REMAINING_FRACTION);
}

/** A worker currently executing a job, as a candidate to WAIT for -- ports
 * `scheduler.BusyWorker`. */
export interface BusyWorker {
  workerId: string;
  name: string;
  backend: string;
  /** Expected seconds until its current job finishes (`remainingSeconds`). */
  etaSeconds: number;
  /** required_models of the job it is running: those will be warm. */
  warmModels: string[];
  inventory: ModelInventoryEntry[];
}

/** Ports `scheduler.BusyWorker.as_worker`. */
export function busyAsWorker(busy: BusyWorker): WorkerCandidate {
  return {
    workerId: busy.workerId,
    name: busy.name,
    backend: busy.backend,
    freeVramGb: 0,
    warmModels: busy.warmModels,
    inventory: busy.inventory,
  };
}

/** Why a job was kept queued this tick: `workerId` is expected to finish it
 * at `waitSeconds` from now, the best idle worker only at `runNowSeconds`
 * (Infinity when no idle worker was eligible at all) -- ports
 * `scheduler.Hold`. */
export interface Hold {
  jobId: string;
  workerId: string;
  waitSeconds: number;
  runNowSeconds: number;
}

/** §3.2 -- ports `scheduler.wait_cost`: expected completion if `job` waits
 * for `busy` -- its remaining time, plus everything already promised to it
 * this tick, plus the same `cost` an idle worker would be charged.
 * `tier1Exists=false`: a busy worker that must fetch is still a real
 * option, the fetch time is in. */
export function waitCost(
  job: JobCandidate,
  busy: BusyWorker,
  pair: PairVerdict,
  predictedSeconds: number,
  queueAheadSeconds: number
): number {
  const base = cost(job, busyAsWorker(busy), pair, predictedSeconds, false);
  if (!finite(base)) return Infinity;
  const total = busy.etaSeconds + queueAheadSeconds + base;
  return finite(total) && total >= 0 ? total : Infinity;
}

/**
 * §3.3 -- ports `scheduler.apply_holds`: decide, job by job in queue order,
 * whether waiting for a busy worker beats every idle worker that could run
 * it now.
 *
 * Returns a copy of `pairs` in which each (job, idle worker) cell that lost
 * to waiting is marked `HELD_KIND` (so `cost` treats it as unavailable),
 * plus one `Hold` per job that ended up with no idle option left. Jobs are
 * walked oldest first and each held job is added to its chosen busy
 * worker's `queueAhead`, so the fifth job waiting for the same card sees
 * the four in front of it and may well go to the Mac after all.
 *
 * Never strands a job: a hold only exists relative to a busy worker that
 * is eligible and has a known ETA. With no busy workers (`busyWorkers`
 * empty) this is the identity, which is also what every caller that
 * predates the feature gets.
 */
export function applyHolds(
  jobs: JobCandidate[],
  workers: WorkerCandidate[],
  busyWorkers: BusyWorker[],
  pairs: Map<string, PairVerdict>,
  predictions: Map<string, number>
): { pairs: Map<string, PairVerdict>; holds: Hold[] } {
  const out = new Map(pairs);
  const holds: Hold[] = [];
  if (busyWorkers.length === 0) return { pairs: out, holds };

  const tier1 = new Map<string, boolean>();
  for (const job of jobs) {
    tier1.set(
      job.jobId,
      workers.some((w) => pairs.get(pairKey(job.jobId, w.workerId))?.kind === "eligible")
    );
  }
  const queueAhead = new Map<string, number>(busyWorkers.map((b) => [b.workerId, 0]));

  for (const job of jobs) {
    let bestWait = Infinity;
    let bestBusy: BusyWorker | null = null;
    for (const busy of busyWorkers) {
      const pair = pairs.get(pairKey(job.jobId, busy.workerId));
      if (!pair) continue;
      const predicted = predictions.get(pairKey(job.jobId, busy.workerId)) ?? 60;
      const value = waitCost(job, busy, pair, predicted, queueAhead.get(busy.workerId)!);
      if (value < bestWait) {
        bestWait = value;
        bestBusy = busy;
      }
    }
    if (bestBusy === null) continue;

    let runNow = Infinity;
    let finiteBefore = 0;
    for (const worker of workers) {
      const key = pairKey(job.jobId, worker.workerId);
      const pair = pairs.get(key);
      if (!pair) continue;
      const predicted = predictions.get(key) ?? 60;
      const value = cost(job, worker, pair, predicted, tier1.get(job.jobId)!);
      if (!finite(value)) continue;
      finiteBefore += 1;
      runNow = Math.min(runNow, value);
      if (value > bestWait * HOLD_MARGIN_RATIO + HOLD_MARGIN_SECONDS) {
        out.set(key, { kind: HELD_KIND, hasWarnings: pair.hasWarnings, totalFetchBytes: pair.totalFetchBytes });
      }
    }

    const stillRunnable = workers.some((w) => {
      const key = pairKey(job.jobId, w.workerId);
      const pair = out.get(key);
      if (!pair) return false;
      return finite(cost(job, w, pair, predictions.get(key) ?? 60, tier1.get(job.jobId)!));
    });
    if (stillRunnable) continue;
    // Waiting (or: nothing idle could run it anyway). Either way this job
    // is next in line on `bestBusy`, so later jobs must queue behind it.
    const own = predictions.get(pairKey(job.jobId, bestBusy.workerId)) ?? 60;
    queueAhead.set(
      bestBusy.workerId,
      queueAhead.get(bestBusy.workerId)! + own + loadSeconds(job, busyAsWorker(bestBusy))
    );
    if (finiteBefore > 0) {
      holds.push({ jobId: job.jobId, workerId: bestBusy.workerId, waitSeconds: bestWait, runNowSeconds: runNow });
    }
  }
  return { pairs: out, holds };
}

/** Ports `scheduler.build_matrix`. */
export function buildMatrix(
  jobs: JobCandidate[],
  workers: WorkerCandidate[],
  pairs: Map<string, PairVerdict>,
  predictions: Map<string, number>,
  now: Date
): number[][] {
  const tier1 = new Map<string, boolean>();
  for (const job of jobs) {
    tier1.set(
      job.jobId,
      workers.some((w) => pairs.get(pairKey(job.jobId, w.workerId))?.kind === "eligible")
    );
  }

  return jobs.map((job) => {
    const waitSeconds = (now.getTime() - job.createdAt.getTime()) / 1000;
    return workers.map((worker) => {
      const pair = pairs.get(pairKey(job.jobId, worker.workerId));
      if (!pair) return Infinity;
      const predicted = predictions.get(pairKey(job.jobId, worker.workerId)) ?? 60;
      return objective(cost(job, worker, pair, predicted, tier1.get(job.jobId)!), waitSeconds);
    });
  });
}

/**
 * 最小成本配對（Kuhn–Munkres / Hungarian，O(n³)）-- ports `scheduler.solve`
 * 逐行。非有限值（Infinity / -Infinity / NaN）視為禁止並在回傳前剔除；
 * 有限負值完全合法（目標函數刻意是負的）。
 *
 * 值域正規化：有限格平移成 `v - minV`，禁止格與補位格一律填
 * `M = (n + 1) * range + 1`。真實配對的總成本差距最多 `n * range < M`，所以
 * 最小化總成本會先最大化真實配對數、再最小化真實成本。不能用固定哨兵
 * （`ulp(1e18) = 128`）：可行圖有缺口時哨兵級的 delta 會進 potential，把整個
 * 成本模型量化掉。
 *
 * 決定性：挑 `delta` 用嚴格小於，平手時永遠選欄位索引最小的那個，和 Python
 * 版一致。呼叫端要先把 jobs 依 `(createdAt, jobId)`、workers 依
 * `(name, workerId)` 排好，這個保證才有意義。
 */
export function solve(matrix: number[][]): [number, number][] {
  const rowsN = matrix.length;
  const colsN = matrix.reduce((max, row) => Math.max(max, row.length), 0);
  const n = Math.max(rowsN, colsN);
  if (n === 0) return [];

  let minV = Infinity;
  let maxV = -Infinity;
  for (const row of matrix) {
    for (const value of row) {
      if (!finite(value)) continue;
      if (value < minV) minV = value;
      if (value > maxV) maxV = value;
    }
  }
  if (!finite(minV)) return [];
  const valueRange = Math.max(maxV - minV, 1);
  const bigCell = (n + 1) * valueRange + 1;

  // 1-indexed 工作矩陣。
  const a: number[][] = Array.from({ length: n + 1 }, () => new Array<number>(n + 1).fill(0));
  for (let i = 0; i < n; i++) {
    for (let j = 0; j < n; j++) {
      let value = bigCell;
      if (i < rowsN && j < matrix[i]!.length) {
        const raw = matrix[i]![j]!;
        if (finite(raw)) value = raw - minV;
      }
      a[i + 1]![j + 1] = value;
    }
  }

  const u = new Array<number>(n + 1).fill(0);
  const v = new Array<number>(n + 1).fill(0);
  const p = new Array<number>(n + 1).fill(0);
  const way = new Array<number>(n + 1).fill(0);

  for (let i = 1; i <= n; i++) {
    p[0] = i;
    let j0 = 0;
    const minv = new Array<number>(n + 1).fill(Infinity);
    const used = new Array<boolean>(n + 1).fill(false);
    for (;;) {
      used[j0] = true;
      const i0 = p[j0]!;
      let delta = Infinity;
      let j1 = 0;
      for (let j = 1; j <= n; j++) {
        if (used[j]) continue;
        const cur = a[i0]![j]! - u[i0]! - v[j]!;
        if (cur < minv[j]!) {
          minv[j] = cur;
          way[j] = j0;
        }
        if (minv[j]! < delta) {
          delta = minv[j]!;
          j1 = j;
        }
      }
      for (let j = 0; j <= n; j++) {
        if (used[j]) {
          u[p[j]!] = u[p[j]!]! + delta;
          v[j] = v[j]! - delta;
        } else {
          minv[j] = minv[j]! - delta;
        }
      }
      j0 = j1;
      if (p[j0] === 0) break;
    }
    for (;;) {
      const j1 = way[j0]!;
      p[j0] = p[j1]!;
      j0 = j1;
      if (j0 === 0) break;
    }
  }

  const result: [number, number][] = [];
  for (let j = 1; j <= n; j++) {
    const i = p[j]!;
    if (i === 0) continue;
    const row = i - 1;
    const col = j - 1;
    if (row >= rowsN || col >= colsN || col >= matrix[row]!.length) continue;
    if (!finite(matrix[row]![col]!)) continue;
    result.push([row, col]);
  }
  result.sort((x, y) => x[0] - y[0] || x[1] - y[1]);
  return result;
}

/** `buildMatrix` + `solve` -- ports `scheduler.match`. */
export function match(
  jobs: JobCandidate[],
  workers: WorkerCandidate[],
  pairs: Map<string, PairVerdict>,
  predictions: Map<string, number>,
  now: Date
): [number, number][] {
  return solve(buildMatrix(jobs, workers, pairs, predictions, now));
}
